docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,529 @@
+
+  def apply_param_gradient(self, step, hyper_params, param, state, grad):
+    del step
+    assert hyper_params.learning_rate is not None
+    new_momentum = state.momentum * hyper_params.beta + grad
+    new_params = param - hyper_params.learning_rate * new_momentum
+    return new_params, _MomentumParamState(new_momentum)
+```
+
+Remarks:
+
+- Note the relationship between `OptimizerDef` and `Optimizer` : When the
+  function `Optimizer.apply_gradient()` is called from the user code, it calls
+  into `OptimizerDef.apply_gradient()` (among other things) which in turn will
+  call `OptimizerDef.apply_param_gradient()` which is implemented by subclasses
+  of `OptimizerDef`.
+- The functions `init_param_state()` and `apply_param_gradient()` are called
+  for every leaf in the params/grads pytree. This makes it possible to write the
+  calculations directly without `jax.tree_multimap()`.
+- The interface was defined in pre-Linen without the distinction of `params` vs.
+  other collections in `variables` in mind. The original API was elegant because
+  one only needed to pass around the optimizer, which included the parameters,
+  optimizer state, optimizer hyperparameters, and a reference to the
+  `OptimizerDef` to perform the param/state update.
+
+## Previous Usage
+[Previous Usage]: #previous-usage
+
+An optimizer would first be constructed from its definition and the pytree of
+target params:
+
+```python
+optimizer_def = flax.optim.Momentum(learning_rate=0.1, beta=0.9)
+optimizer = optimizer_def.create(variables['params'])
+```
+
+The target variables would then optimized in the train step (assuming a single
+non-params collection "batch_stats"):
+
+```python
+def make_train_step(apply_fn):
+  @jax.jit
+  def train_step(optimizer, batch_stats, inputs, labels):
+
+    def loss_fn(params):
+      variables = {'params': params, 'batch_stats': batch_stats}
+      logits, new_model_state = apply_fn(
+          variables, inputs, mutable=['batch_stats'])
+      loss = xent_loss(logits, labels)
+      return loss, new_model_state['batch_stats']
+
+    (loss, new_batch_stats), grad = jax.value_and_grad(loss_fn, has_aux=True)(
+        optimizer.target)
+    lr = get_learning_rate(step)
+    new_optimizer = optimizer.apply_gradient(grad, learning_rate=lr)
+    return new_optimizer, new_batch_stats, loss
+
+  return train_step
+
+
+batch_stats = variables['batch_stats']
+train_step = make_train_step(model.apply)
+for step, batch in enumerate(ds)
+  optimizer, batch_stats, loss = train_step(
+      optimizer, batch_stats, batch['image'], batch['label'])
+```
+
+Remarks:
+
+- Note how `optimizer.apply_gradients()` can take additional arguments to update
+  hyper parameters, like the learning rate from an independent function
+  `get_learning_rate()` in this case.
+
+
+# Update Plan
+[Update Plan]: #update-plan
+
+1. Finalize discussions on this FLIP
+2. Test existing optimizers for numerical equivalence (e.g. `flax.optim.Adam`
```

Comment:
I believe @mtthss was interested in helping with this -- it would entail copying over the Flax optimizers into the Optax test directory as the hope is to eventually deprecate flax.optimizers so we shouldn't do this by having optax tests depend on Flax itself.

Reply:
I think that's a good idea. If somebody from Optax team can confirm here, then I'll update the description accordingly.

Reply:
Yes I am happy to help with this!